src/zx/FunctionalityConstruction.cpp

```diff
@@ -0,0 +1,344 @@
+        addZSpider(diag, ctrl, qubit_vertices);
+        addXSpider(diag, target, qubit_vertices);
+        diag.addEdge(qubit_vertices[ctrl], qubit_vertices[target]);
+    }
+
+    void FunctionalityConstruction::addCphase(ZXDiagram& diag, const PiRational& phase, Qubit ctrl, Qubit target,
+                                              std::vector<Vertex>& qubit_vertices) {
+        addZSpider(diag, ctrl, qubit_vertices, Expression(phase / 2));
+        addCnot(diag, ctrl, target, qubit_vertices);
+        addZSpider(diag, target, qubit_vertices, Expression(-phase / 2));
+        addCnot(diag, ctrl, target, qubit_vertices);
+        addZSpider(diag, target, qubit_vertices, Expression(phase / 2));
+    }
+
+    void FunctionalityConstruction::addSwap(ZXDiagram& diag, Qubit ctrl, Qubit target,
+                                            std::vector<Vertex>& qubit_vertices) {
+        const auto s0 = qubit_vertices[target];
+        const auto s1 = qubit_vertices[ctrl];
+
+        const auto t0 = diag.addVertex(target, diag.getVData(qubit_vertices[target]).value().col + 1);
+        const auto t1 = diag.addVertex(ctrl, diag.getVData(qubit_vertices[target]).value().col + 1);
+
+        diag.addEdge(s0, t1);
+        diag.addEdge(s1, t0);
+
+        qubit_vertices[target] = t0;
+        qubit_vertices[ctrl]   = t1;
+    }
+
+    void FunctionalityConstruction::addCcx(ZXDiagram& diag, Qubit ctrl_0, Qubit ctrl_1, Qubit target,
+                                           std::vector<Vertex>& qubit_vertices) {
+        addZSpider(diag, target, qubit_vertices, Expression(), EdgeType::Hadamard);
+        addCnot(diag, ctrl_1, target, qubit_vertices);
+        addZSpider(diag, target, qubit_vertices, Expression(PiRational(-1, 4)));
+        addCnot(diag, ctrl_0, target, qubit_vertices);
+        addZSpider(diag, target, qubit_vertices, Expression(PiRational(1, 4)));
+        addCnot(diag, ctrl_1, target, qubit_vertices);
+        addZSpider(diag, ctrl_1, qubit_vertices, Expression(PiRational(1, 4)));
+        addZSpider(diag, target, qubit_vertices, Expression(PiRational(-1, 4)));
+        addCnot(diag, ctrl_0, target, qubit_vertices);
+        addZSpider(diag, target, qubit_vertices, Expression(PiRational(1, 4)));
+        addCnot(diag, ctrl_0, ctrl_1, qubit_vertices);
+        addZSpider(diag, ctrl_0, qubit_vertices, Expression(PiRational(1, 4)));
+        addZSpider(diag, ctrl_1, qubit_vertices, Expression(PiRational(-1, 4)));
+        addZSpider(diag, target, qubit_vertices, Expression(PiRational(0, 1)),
+                   EdgeType::Hadamard);
+        addCnot(diag, ctrl_0, ctrl_1, qubit_vertices);
+    }
+
+    FunctionalityConstruction::op_it FunctionalityConstruction::parse_op(ZXDiagram& diag, op_it it, op_it end,
+                                                                         std::vector<Vertex>& qubit_vertices) {
+        auto& op = *it;
+
+        if (op->getType() == qc::OpType::Barrier) {
+            return it + 1;
+        }
```

Comment:
these lines should be easy to cover in a test (probably even in an existing one).